Tools/CISupport/Shared/de-duper

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python3
+# Copyright (C) 2024 Apple Inc. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1.  Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+# 2.  Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import argparse
+import hashlib
+import os
+import sys
+
+CHUNK_SIZE = 8192
+
+
+def parse_args(args):
+    parser = argparse.ArgumentParser(
+        description='Program which de-duplicates PNG files in a directory using hard links'
+    )
+    parser.add_argument(
+        'root', nargs='?',
+        help='Path to root of directory to de-duplicate PNG files in',
+        type=str, default='.',
+    )
+    parser.add_argument(
+        '-s', '--storage',
+        help='Directory to store images the program is linking to',
+        type=str,
+        default='storage',
+    )
+    parser.add_argument(
+        '-d', '--dry-run',
+        action='store_true',
+        dest='dry_run',
+        default=False,
+    )
+    return parser.parse_args(args)
+
+
+def sha1(path):
+    try:
+        with open(path, 'rb') as file:
+            hsh = hashlib.sha1()
+            chunk = file.read(CHUNK_SIZE)
+            while chunk:
+                hsh.update(chunk)
+                chunk = file.read(CHUNK_SIZE)
+        return hsh.hexdigest()
+    except OSError:
+        return None
+
+
+def main(args):
+    args = parse_args(args)
+
+    args.root = os.path.abspath(args.root)
+    args.storage = os.path.abspath(args.storage) if args.storage[0] == '/' else os.path.join(args.root, args.storage)
```

Comment:
There's no need to special-case `args.storage[0] == '/'`; `os.path.join` does the right thing itself.